aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -348,51 +794,90 @@ module aptos_framework::coin {
         // Skip burning if amount is zero. This shouldn't error out as it's called as part of transaction fee burning.
         if (amount == 0) {
             return
         };
 
-        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
-        let coin_to_burn = extract(&mut coin_store.coin, amount);
-        burn(coin_to_burn, burn_cap);
+        let (coin_amount_to_burn, fa_amount_to_burn) = calculate_amount_to_withdraw<CoinType>(
+            account_addr,
+            amount
+        );
+        if (coin_amount_to_burn > 0) {
+            let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
+            let coin_to_burn = extract(&mut coin_store.coin, coin_amount_to_burn);
+            burn(coin_to_burn, burn_cap);
+        };
+        if (fa_amount_to_burn > 0) {
+            fungible_asset::burn_from(
+                borrow_paired_burn_ref(burn_cap),
+                primary_fungible_store::primary_store(account_addr, option::destroy_some(paired_metadata<CoinType>())),
+                fa_amount_to_burn
+            );
+        };
     }
 
     /// Deposit the coin balance into the recipient's account and emit an event.
-    public fun deposit<CoinType>(account_addr: address, coin: Coin<CoinType>) acquires CoinStore {
-        assert!(
-            is_account_registered<CoinType>(account_addr),
-            error::not_found(ECOIN_STORE_NOT_PUBLISHED),
-        );
-
-        let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
-        assert!(
-            !coin_store.frozen,
-            error::permission_denied(EFROZEN),
-        );
-        if (std::features::module_event_migration_enabled()) {
-            event::emit(Deposit<CoinType> { account: account_addr, amount: coin.value });
-        };
-        event::emit_event<DepositEvent>(
-            &mut coin_store.deposit_events,
-            DepositEvent { amount: coin.value },
-        );
+    public fun deposit<CoinType>(
+        account_addr: address,
+        coin: Coin<CoinType>
+    ) acquires CoinStore, CoinConversionMap, CoinInfo {
+        if (exists<CoinStore<CoinType>>(account_addr)) {
+            let coin_store = borrow_global_mut<CoinStore<CoinType>>(account_addr);
+            assert!(
+                !coin_store.frozen,
+                error::permission_denied(EFROZEN),
+            );
+            if (std::features::module_event_migration_enabled()) {
+                event::emit(Deposit<CoinType> { account: account_addr, amount: coin.value });
+            };
+            event::emit_event<DepositEvent>(
+                &mut coin_store.deposit_events,
+                DepositEvent { amount: coin.value },
+            );
+            merge(&mut coin_store.coin, coin);
+        } else {
+            if (features::coin_to_fungible_asset_migration_feature_enabled(
+            ) && migrated_primary_fungible_store_exists(account_addr, ensure_paired_metadata<CoinType>())) {
```

Comment:
newly added APT gating will not work correctly for this, I think.

one simplification question, why do we auto-upgrade on calls, instead of just checking if mapping exists, and then just have an entry function, that anyone can call, that can create a mapping.

otherwise we probably need to have try_paired_metadata, to try to pair them, and keep the feature gating checks inside.

Reply:
Now the code will error out with the new gate but not technically incorrect. I agree it is not ideal.

I will refactor this part to not create fungible store if does not exist so use `paired_metadata` but not `ensure_paired_metadata` then we should be good.

manual upgrade introduces one more steps and make the migration a little bit harder. From my understanding, adding any more step would prohibit the migration so let's minimize the work of the end users as much as possible if it is not cumbersome or complicated on our side.